### PR TITLE
fix nested select when used inside a nested form

### PR DESCRIPTION
**Pull request by rubenbe**
Fixed style issues of #164. Eslint is happy now

There are currently no nested forms in the test setup that can be used to test this.
I could add a new "citygroup" that contains multiple cities and uses a nested form

### CODE REVIEW

**Comment by rubenbe** [comment]
@ldlsegovia is it sufficient if I add the test cases as described above, to have this fix (considered to be) merged?
If there are extra requirements, please let me know.

**Comment by ldlsegovia** [comment]
@rubenbe yo need to add your specs here: https://github.com/platanus/activeadmin_addons/blob/master/spec/features/inputs/nested_select_input_spec.rb
You could add a new context `"with nested resources"` and render a `Region` form with a nested form to add new cities using the `register_form` method.

File: spec/support/capybara_helpers.rb [hunk 1]
```diff
@@ -95,4 +95,20 @@ def expect_select2_results_count_to_eq(count)
       expect(page).not_to have_content(no_results)
     end
   end
+
+  def click_add_nested()
+    find("a.has_many_add").click
+  end
+
+  def on_nested_ctx(resource_number, &block)
+    within("li.has_many_container fieldset:nth-child(#{resource_number+1}) ") do
```
Comment (monkeyci) [inline spec/support/capybara_helpers.rb]: Surrounding space missing for operator +.

File: spec/support/capybara_helpers.rb [hunk 2]
```diff
@@ -95,4 +95,20 @@ def expect_select2_results_count_to_eq(count)
       expect(page).not_to have_content(no_results)
     end
   end
+
+  def click_add_nested()
```
Comment (monkeyci) [inline spec/support/capybara_helpers.rb]: Omit the parentheses in defs when the method doesn't accept any arguments.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 3]
```diff
@@ -185,4 +185,66 @@
...
+      on_nested_ctx(2) do
+        on_input_ctx("#{prefix}1_country_id"){
+          open_select2_options
+          click_select2_option(@argentina.id)
+        }
+        on_input_ctx("#{prefix}1_region_id"){
+          open_select2_options
+          click_select2_option(@cuyo.id)
+        }
+        on_input_ctx("#{prefix}1_city_id"){
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Space missing to the left of {.<br>Avoid using {...} for multi-line blocks.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 4]
```diff
@@ -185,4 +185,66 @@
...
+        expect_nested_select2_result_text_to_eq(3, @santiago.id)
+      end
+
+      click_add_nested
+      on_nested_ctx(2) do
+        on_input_ctx("#{prefix}1_country_id"){
+          open_select2_options
+          click_select2_option(@argentina.id)
+        }
+        on_input_ctx("#{prefix}1_region_id"){
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Space missing to the left of {.<br>Avoid using {...} for multi-line blocks.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 5]
```diff
@@ -185,4 +185,66 @@
...
+        on_input_ctx("#{prefix}0_city_id"){
+          open_select2_options
+          click_select2_option(@santiago.id)
+        }
+        expect_nested_select2_result_text_to_eq(3, @santiago.id)
+      end
+
+      click_add_nested
+      on_nested_ctx(2) do
+        on_input_ctx("#{prefix}1_country_id"){
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Space missing to the left of {.<br>Avoid using {...} for multi-line blocks.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 6]
```diff
@@ -185,4 +185,66 @@
...
+      on_nested_ctx(1) do
+        on_input_ctx("#{prefix}0_country_id"){
+          open_select2_options
+          click_select2_option(@chile.id)
+        }
+        on_input_ctx("#{prefix}0_region_id"){
+          open_select2_options
+          click_select2_option(@metropolitana.id)
+        }
+        on_input_ctx("#{prefix}0_city_id"){
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Space missing to the left of {.<br>Avoid using {...} for multi-line blocks.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 7]
```diff
@@ -185,4 +185,66 @@
...
+
+    it "sets new values correctly", js: true do
+      click_add_nested
+      prefix = "department_departments_cities_attributes_"
+      on_nested_ctx(1) do
+        on_input_ctx("#{prefix}0_country_id"){
+          open_select2_options
+          click_select2_option(@chile.id)
+        }
+        on_input_ctx("#{prefix}0_region_id"){
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Space missing to the left of {.<br>Avoid using {...} for multi-line blocks.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 8]
```diff
@@ -185,4 +185,66 @@
...
+      end
+      create_cities
+      visit new_admin_department_path
+    end
+
+    it "sets new values correctly", js: true do
+      click_add_nested
+      prefix = "department_departments_cities_attributes_"
+      on_nested_ctx(1) do
+        on_input_ctx("#{prefix}0_country_id"){
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Space missing to the left of {.<br>Avoid using {...} for multi-line blocks.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 9]
```diff
@@ -185,4 +185,66 @@
...
+      register_page(City, false) {}
+
+      register_form(Department, false) do |f|
+        f.has_many :departments_cities, allow_destroy: true do |city|
+          city.input :city_id, as: :nested_select, :required => true,
+                display_name: :id,
+                minimum_input_length: 0,
+                level_1: { attribute: :country_id },
+                level_2: { attribute: :region_id },
+                level_3: { attribute: :city_id }
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Align the elements of a hash literal if they span more than one line.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 10]
```diff
@@ -185,4 +185,66 @@
...
+      register_page(Country, false) {}
+      register_page(City, false) {}
+
+      register_form(Department, false) do |f|
+        f.has_many :departments_cities, allow_destroy: true do |city|
+          city.input :city_id, as: :nested_select, :required => true,
+                display_name: :id,
+                minimum_input_length: 0,
+                level_1: { attribute: :country_id },
+                level_2: { attribute: :region_id },
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Align the elements of a hash literal if they span more than one line.

**Comment by rubenbe** [comment]
I added a department model which can refer to multiple cities. I required a many to many relationship to a nested resource which was not immediately available

File: spec/features/inputs/nested_select_input_spec.rb [hunk 11]
```diff
@@ -185,4 +185,66 @@
...
+
+  context "with nested resources" do
+    before do
+      register_page(Region, false) {}
+      register_page(Country, false) {}
+      register_page(City, false) {}
+
+      register_form(Department, false) do |f|
+        f.has_many :departments_cities, allow_destroy: true do |city|
+          city.input :city_id, as: :nested_select, :required => true,
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Use the new Ruby 1.9 hash syntax.

File: spec/dummy/db/schema.rb [hunk 12]
```diff
@@ -76,6 +76,20 @@
...
+    t.datetime "updated_at", null: false
+  end
+
+  create_table "departments_cities", force: true do |t|
+    t.integer "department_id"
+    t.integer "city_id"
+  end
+
+  add_index "departments_cities", ["city_id"], name: "index_departments_cities_on_city_id"
+  add_index "departments_cities", ["department_id"], name: "index_departments_cities_on_department_id"
```
Comment (monkeyci) [inline spec/dummy/db/schema.rb]: Line is too long. [102/100]

File: spec/dummy/app/admin/departments.rb [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+  form do |f|
+    f.inputs do
+      f.input :name
+    end
+    f.inputs do
+      f.has_many :departments_cities, allow_destroy: true do |city|
+        city.input :city_id, as: :nested_select, :required => true,
+              level_1: { attribute: :country_id },
+              level_2: { attribute: :region_id },
+              level_3: { attribute: :city_id }
```
Comment (monkeyci) [inline spec/dummy/app/admin/departments.rb]: Align the elements of a hash literal if they span more than one line.

File: spec/dummy/app/admin/departments.rb [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+  form do |f|
+    f.inputs do
+      f.input :name
+    end
+    f.inputs do
+      f.has_many :departments_cities, allow_destroy: true do |city|
+        city.input :city_id, as: :nested_select, :required => true,
+              level_1: { attribute: :country_id },
+              level_2: { attribute: :region_id },
```
Comment (monkeyci) [inline spec/dummy/app/admin/departments.rb]: Align the elements of a hash literal if they span more than one line.

File: spec/dummy/app/admin/departments.rb [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+  permit_params :name, departments_cities_attributes: [:id, :city_id, :_destroy]
+
+  form do |f|
+    f.inputs do
+      f.input :name
+    end
+    f.inputs do
+      f.has_many :departments_cities, allow_destroy: true do |city|
+        city.input :city_id, as: :nested_select, :required => true,
+              level_1: { attribute: :country_id },
```
Comment (monkeyci) [inline spec/dummy/app/admin/departments.rb]: Align the elements of a hash literal if they span more than one line.

File: spec/dummy/app/admin/departments.rb [hunk 16]
```diff
@@ -0,0 +1,18 @@
+ActiveAdmin.register Department do
+  permit_params :name, departments_cities_attributes: [:id, :city_id, :_destroy]
+
+  form do |f|
+    f.inputs do
+      f.input :name
+    end
+    f.inputs do
+      f.has_many :departments_cities, allow_destroy: true do |city|
+        city.input :city_id, as: :nested_select, :required => true,
```
Comment (monkeyci) [inline spec/dummy/app/admin/departments.rb]: Use the new Ruby 1.9 hash syntax.

File: spec/dummy/db/schema.rb (hunk 12 above)
Comment (rubenbe) [reply spec/dummy/db/schema.rb]: I asked rubocop to ignore this file

**Comment by rubenbe** [comment]
Hi @ldlsegovia,
Do you need more testcase to have this merged?

**Comment by ldlsegovia** [comment]
@rubenbe it looks good! a final request... Could you remove the `DepartmentsCity` model and use a [HABTM association](http://guides.rubyonrails.org/association_basics.html#the-has-and-belongs-to-many-association)? you don't really need that model.

And please, do a rebase to resolve the conflicts.

**Comment by rubenbe** [comment]
@ldlsegovia in that case I got an error "region_id is already defined" which is probably due to the "belongs_to :region" in the City model clashing with functions that ActiveAdmin wants to define.
I'll look into it a bit further.

**Comment by jgmontoya** [comment]
I think this is long overdue, we should push this forward. From the different PR's addressing this issue this one seems to be the more complete one.

@rubenbe are you still interested in this? If so please rebase and I will make sure this doesn't get forgotten.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 17]
```diff
@@ -185,4 +185,66 @@
...
+        end
+        on_input_ctx("#{prefix}1_city_id") do
+          open_select2_options
+          click_select2_option(@mendoza.id)
+        end
+        expect_nested_select2_result_text_to_eq(3, @mendoza.id)
+      end
+
+      on_nested_ctx(1) do
+        expect_nested_select2_result_text_to_eq(3, @santiago.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 18]
```diff
@@ -185,4 +185,66 @@
...
+        end
+        on_input_ctx("#{prefix}1_region_id") do
+          open_select2_options
+          click_select2_option(@cuyo.id)
+        end
+        on_input_ctx("#{prefix}1_city_id") do
+          open_select2_options
+          click_select2_option(@mendoza.id)
+        end
+        expect_nested_select2_result_text_to_eq(3, @mendoza.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 19]
```diff
@@ -185,4 +185,66 @@
...
+          open_select2_options
+          click_select2_option(@argentina.id)
+        end
+        on_input_ctx("#{prefix}1_region_id") do
+          open_select2_options
+          click_select2_option(@cuyo.id)
+        end
+        on_input_ctx("#{prefix}1_city_id") do
+          open_select2_options
+          click_select2_option(@mendoza.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 20]
```diff
@@ -185,4 +185,66 @@
...
+
+      click_add_nested
+      on_nested_ctx(2) do
+        on_input_ctx("#{prefix}1_country_id") do
+          open_select2_options
+          click_select2_option(@argentina.id)
+        end
+        on_input_ctx("#{prefix}1_region_id") do
+          open_select2_options
+          click_select2_option(@cuyo.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 21]
```diff
@@ -185,4 +185,66 @@
...
+          click_select2_option(@santiago.id)
+        end
+        expect_nested_select2_result_text_to_eq(3, @santiago.id)
+      end
+
+      click_add_nested
+      on_nested_ctx(2) do
+        on_input_ctx("#{prefix}1_country_id") do
+          open_select2_options
+          click_select2_option(@argentina.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 22]
```diff
@@ -185,4 +185,66 @@
...
+        end
+        on_input_ctx("#{prefix}0_region_id") do
+          open_select2_options
+          click_select2_option(@metropolitana.id)
+        end
+        on_input_ctx("#{prefix}0_city_id") do
+          open_select2_options
+          click_select2_option(@santiago.id)
+        end
+        expect_nested_select2_result_text_to_eq(3, @santiago.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 23]
```diff
@@ -185,4 +185,66 @@
...
+          open_select2_options
+          click_select2_option(@chile.id)
+        end
+        on_input_ctx("#{prefix}0_region_id") do
+          open_select2_options
+          click_select2_option(@metropolitana.id)
+        end
+        on_input_ctx("#{prefix}0_city_id") do
+          open_select2_options
+          click_select2_option(@santiago.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 24]
```diff
@@ -185,4 +185,66 @@
...
+      click_add_nested
+      prefix = "department_departments_cities_attributes_"
+      on_nested_ctx(1) do
+        on_input_ctx("#{prefix}0_country_id") do
+          open_select2_options
+          click_select2_option(@chile.id)
+        end
+        on_input_ctx("#{prefix}0_region_id") do
+          open_select2_options
+          click_select2_option(@metropolitana.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 25]
```diff
@@ -185,4 +185,66 @@
...
+      visit new_admin_department_path
+    end
+
+    it "sets new values correctly", js: true do
+      click_add_nested
+      prefix = "department_departments_cities_attributes_"
+      on_nested_ctx(1) do
+        on_input_ctx("#{prefix}0_country_id") do
+          open_select2_options
+          click_select2_option(@chile.id)
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 26]
```diff
@@ -185,4 +185,66 @@
...
+                               level_1: { attribute: :country_id },
+                               level_2: { attribute: :region_id },
+                               level_3: { attribute: :city_id }
+        end
+      end
+      create_cities
+      visit new_admin_department_path
+    end
+
+    it "sets new values correctly", js: true do
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Example has too many lines [36/10].

**Comment by rubenbe** [comment]
@jgmontoya rebased

File: .rubocop.yml [hunk 27]
```diff
@@ -1,6 +1,7 @@
 AllCops:
   Exclude:
   - "vendor/**/*"
+  - "db/*"
```
Comment (jgmontoya) [inline .rubocop.yml]: I might be wrong here but I think this should be `spec/dummy/db/*`

File: spec/dummy/db/schema.rb (hunk 12 above)
Comment (jgmontoya) [reply spec/dummy/db/schema.rb]: I think you added the wrong path for rubocop to ignore, see the other comment!

File: spec/features/inputs/nested_select_input_spec.rb [hunk 28]
```diff
@@ -185,4 +185,68 @@
...
+        on_input_ctx("#{prefix}0_city_id") do
+          open_select2_options
+          click_select2_option(santiago)
+        end
+        expect_nested_select2_result_text_to_eq(3, santiago)
+      end
+
+      click_add_nested
+      on_nested_ctx(2) do
+        mendoza = @mendoza.id
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 29]
```diff
@@ -185,4 +185,68 @@
...
+        end
+      end
+      create_cities
+      visit new_admin_department_path
+    end
+
+    it "sets new values correctly", js: true do
+      click_add_nested
+      prefix = "department_departments_cities_attributes_"
+      santiago = @santiago.id
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Replace instance variable with local variable or let.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 30]
```diff
@@ -185,4 +185,68 @@
...
+                               level_1: { attribute: :country_id },
+                               level_2: { attribute: :region_id },
+                               level_3: { attribute: :city_id }
+        end
+      end
+      create_cities
+      visit new_admin_department_path
+    end
+
+    it "sets new values correctly", js: true do
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb]: Example has too many lines [38/10].

File: spec/dummy/app/models/departments_city.rb [hunk 31]
```diff
@@ -0,0 +1,4 @@
+class DepartmentsCity < ActiveRecord::Base
```
Comment (monkeyci) [inline spec/dummy/app/models/departments_city.rb]: Models should subclass ApplicationRecord.

File: spec/dummy/app/models/department.rb [hunk 32]
```diff
@@ -0,0 +1,4 @@
+class Department < ActiveRecord::Base
```
Comment (monkeyci) [inline spec/dummy/app/models/department.rb]: Models should subclass ApplicationRecord.

**Comment by jgmontoya** [comment]
Sorry about that!
Please revert back to `ActiveRecord::Base` as the change to `ApplicationRecord` would require us to upgrade from rails 4 to rails 5.

**Comment by rubenbe** [comment]
Thanks for merging!

**Pull request merged**